### ISSUE TITLE
Linux support (WebSocket client)

### ISSUE THREAD

**Issue by pixelspark** [daltoniam/Starscream#346]
Would really like to have Starscream on Linux as WebSocket client (a WebSocket server is already available through Kitura-WebSocket but a WebSocket client is currently missing, and the server WG has not yet produced anything usable at this point as far as I know).

Would it be feasible to switch over to a generic socket library (such as https://github.com/IBM-Swift/BlueSocket ? I believe there is also SSL support in there for both macOS and Linux).

#285 discussed Linux support before, but that concerned a WebSocket _server_.

**Comment by daltoniam** [comment]
I would like to see Linux support too 😄. I think it might be feasible, I know in #335 there has been talk of allowing a swap out of different socket connections/adapters which would probably be the most flexible route. It is probably worth looking at with the Xcode 9 and Swift 4 update that will be need here in the near future.

**Comment by daltoniam** [comment]
I did some work with the 3.0.0 release to allow this. There is a new `WSStream` class that can be implemented to allow a custom socket to be used. I also add some os checks for linux around the security framework since that doesn't exist on non apple platforms. I haven't done a full linux test yet, but I'm hopeful it should be feasible now.

**Comment by pixelspark** [comment]
@daltoniam cool stuff! Will have a look at it later on. How is/will security processing (e.g. the hashing required to establish a connection) (be) done on Linux? I suppose you need to use CommonCrypto on OSX/iOS and openssl on Linux (perhaps just use a wrapper library such as IBM-Swift/Cryptor, which has worked well for me)?

**Comment by acmacalister** [comment]
@pixelspark, yeah that is what we are probably looking at. We haven't heard or seen any signs of the `Security.framework` making it's way over to Linux, so a wrapper around OpenSSL is probably going to be our best bet. As far as the sha1 hashing is concerned, we could use the OpenSSL version or implement our own in pure Swift like CryptoSwift does. For the certificate/trust validation, I think OpenSSL is going to be the only way. Good catch on sha1 hashing still using CC. Should be fairly easy to move that out of the `WSStream` class and make something flexible for all the platforms.

**Comment by caiofbpa** [comment]
Any way I can help this move forward? I got IBM-Swift/Cryptor included in Starscream locally, but I couldn't manage to run Autobahn|Testsuite to check if it's all good...

**Comment by wbarksdale** [comment]
@caiofbpa I would be interested in testing out this branch!

**Comment by caiofbpa** [comment]
@wbarksdale I've opened a PR with that branch on #454. Can you help me test and review it?

**Comment by c-bata** [comment]
Hi. Is there any update on this?

**Comment by fassko** [comment]
I think this should be solved with SwiftNIO. I have plan to integrate that into Starscream at some point.

**Comment by woodcrust** [comment]
@fassko if to use "Swift-NIO", then I think this library is no longer needed :D

**Comment by fassko** [comment]
It won't because SwiftNIO needs Network.framework which is only for iOS 12+, tvOS 12+, macOS 10.14+ and exclude watchOS.

**Comment by alex-taffe** [comment]
Any updates on this?

**Comment by fassko** [comment]
Not really, maybe you can help out?

**Comment by alex-taffe** [comment]
@fassko do you know if any effort has been on this so far and if so, can you point me in the right direction?

**Comment by nerzh** [comment]
@alex-taffe but why do you need it on Linux ?

**Comment by alex-taffe** [comment]
@nerzh Attempting to build a cross platform socket based framework, could do it in another language but prefer the type safety of Swift and ease of integration in iOS apps

**Comment by fassko** [comment]
@alex-taffe it should work by default, but I don't have Linux box to test on myself.

**Comment by alex-taffe** [comment]
Ok I’ll do some testing, thanks. Looks like Vapor also may be an option but more complicated than I really need

**Comment by nerzh** [comment]
@alex-taffe I am using https://github.com/vapor/websocket-kit and happy. For example you can see code based on websocket-kit under spoiler from readme of this repo https://github.com/nerzh/Action-Cable-Swift
